### PR TITLE
Convert http to https and update 301 forwarded urls

### PR DESCRIPTION
**Pull request by velaja**
this bash script operates various maintenance tasks on files urls and cache.db for newsboat:
 - convert http to https in urls file when https feed is available
 - convert feeds urls to their forwarded urls for 3xx http codes
 - feeds urls not returning a valid http code 200 are tagged "_dead"
 cache.db is also updated to refer to the new url feeds so all read/unread articles and flags are preserved

### CODE REVIEW

File: contrib/urls-maintenance.sh [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+# https://www.linuxjournal.com/content/parsing-rss-news-feed-bash-script
+#	is it returning valid rss?
+#	when was the feed last updated?
+#	sort valid feeds by last updated
+#	tag feed "abandoned" when feed not updated for more than 1 year
+
+
+#newsboat urls file and cache locations
+u="$HOME/.config/newsboat/urls"
+db="$HOME/.config/newsboat/cache.db"
```
Comment (Minoru) [inline contrib/urls-maintenance.sh]: Is this a typo? The cache [should go](https://newsboat.org/releases/2.16.1/docs/newsboat.html#_xdg_base_directory_support) to `~/.local/share/newsboat`.

File: contrib/urls-maintenance.sh [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+#curl timeout for testing URLs
+timeout=10
+tagdead="_dead"
+
+# shuf (GNU coreutils) randomises the urls list, this avoids querying the same domains too fast, assuming urls are grouped by domains or alphasorted in the urls file
+requirements=( newsboat curl sqlite3 sed grep awk head shuf command )
+for app in ${requirements[*]}
```
Comment (Minoru) [inline contrib/urls-maintenance.sh]: Arrays are a bash-ism, please update the shebang to use `env /bin/bash`.

File: contrib/urls-maintenance.sh [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+
+cp "$db" "$db.bak-$(date +%FT%T)"
+cp "$u" "$u.bak-$(date +%FT%T)"
+
+feeds=$(wc -l "$u" | awk '{print $1}')
+i=0
+# check all feeds return valid http codes
+for url in $(shuf "$u" | awk '{print $1}')
+do
+	i=$((i+1))
```
Comment (tsipinakis) [inline contrib/urls-maintenance.sh]: [nitpick] Given we're going full-on bash let's make it prettier. 
```suggestion
	let i=i+1
```

File: contrib/urls-maintenance.sh [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+	#clear the line before echoing over it
+	echo -ne "\r\e[K$i/$feeds $url\r"
+	response=$(curl -I --connect-timeout "$timeout" --write-out "%{http_code}" --silent --output /dev/null "$url")
```
Comment (Minoru) [inline contrib/urls-maintenance.sh]: I think it would be better to use `--max-time` instead of `--connect-timeout` here; during my testing, the script ran into a site that took about a minute to check. I can't reproduce that anymore, but `--max-time` seem like a sure-fire way to avoid this class of situations.

File: contrib/urls-maintenance.sh [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+	case "$response" in
+		3*)
+			url2=$(curl -IL --silent "$url" | awk '/^[lL]ocation: /{ print $2 }' | head -1 | sed 's/\r//g')
```
Comment (Minoru) [inline contrib/urls-maintenance.sh]: It probably makes sense to add `--max-time` here as well. Surely 10 seconds is enough to unravel any possible chain of redirects and arrive at the final destination!

**Comment by Minoru** [comment]
I finally tried this on my urls file, and here's the results.

Most of what the script did is really useful; I'm glad @velaja took the time to write it. I didn't know so many of my feeds migrated to HTTPS already!

There were a couple puzzling situations, though. For example, this:

```
302 [ https://httpstatuses.com/301 ] https://www.theguardian.com/books/rss
                            moved to https://www.theguardian.com/books/rss
```

The HTTP code and httpstatuses URL don't match, and the original and moved-to URLs are the same. I'm not sure if this is indicative of a bug in a script, but I'd like @velaja to check, just in case.

Then there were these:

```
301 [ https://httpstatuses.com/301 ] https://jacobian.org/feed.xml
                            moved to /atom/entries/

301 [ https://httpstatuses.com/301 ] https://www.ft.com/markets/commodities?format=rss
                            moved to /commodities?format=rss

```

Totally innocuous—the URL in the URLs file was updated to "https://jacobian.org/atom/entries". But I'd still like the output of the script to display the full URL as well, because it's alarming to see just the path without the domain.

`exec:` is not supported. I suspect neither is `filter:`. My only `exec:` got all wrecked:

```
# before
"exec:~/.local/bin/github-search-to-rss newsboat"

# after
"exec:~/.local/bin/github-search-to-rss _dead newsboat"
```

Finally, the worst issue of all: harmful redirects. For example:

```
302 [ https://httpstatuses.com/302 ] http://abstrusegoose.com/feed
                            moved to https://abstrusegoose.com/406.shtml
```

The new URL returns 404 to boot, which is terrible. I think this should be mitigated two-fold:
1. do an additional HEAD request to check that the moved-to URL actually returns 2xx code;
2. pass `--user-agent` to CURL, with a user-configurable value as a parameter. (A user agent from recent-ish Firefox ought to be a nice default.) This is needed because that redirect doesn't happen in the browser; apparently that site simply doesn't like CURL.

One more example of a bad redirect:

```
301 [ https://httpstatuses.com/301 ] http://it-the-drote.tk/rss.xml
                            moved to https://blog.rcmd.space
```

The redirect is *kinda* correct: the site moved to a new domain. However, it's a "blanket" redirect that moves everything to the front page. I think the only way to combat those is to check the `Content-Type` header of the response to ensure that we're getting a feed of some sort, not an HTML page.

Ultimately, I think it's better if the script migrates stuff only when it is absolutely sure the migration is correct, and displays warnings when migration seem possible but doesn't look exactly correct (e.g. Content-Type is wrong). This gives the user an option to manually check afterwards.

Finally, please add an execution bit to the script.

Sorry for a really long comment, but real life served some interesting corner cases for the script :) Once more, thanks for working on this, @velaja!

**Comment by velaja** [comment]
thanks everybody for all your detailed feedbacks.
I resolved all the above issues, made the script POSIX-compliant, and I found more corner cases such as https://news.ycombinator.com/rss
this feed will return 405 on HEAD requests so I need to GET the actual feed.
I am doing a few more tests and I will post the update.

**Comment by Minoru** [comment]
Cool! I really like the use of HEAD, so if you can, please keep it and fall back to GET only if code 405 is returned.

File: contrib/urls-maintenance.sh (hunk 5 above)
Comment (velaja) [reply contrib/urls-maintenance.sh]: thanks for the switch recommendation. I raised the default timeout to 20 seconds due to some rss-bridge feeds taking over 10s because they need to pull descriptions from all videos in youtube videos channels feeds.

**Comment by Minoru** [comment]
@velaja, how's those tests going? Still hoping for an update ^_^

**Comment by velaja** [comment]
I commented out the part adding a "_fail" tag for now, it just echoes there may be a problem there. I found some feeds replying 403 but the curl stdout is a correct rss.
I'm using GET instead of HEAD because some feeds don't like HEAD and this is also in preparation for the upcoming feature to check pubdate for the last entry in each feed. This will be useful to find which feeds have not been updated in months or years.

File: contrib/urls-maintenance.sh (hunk 1 above)
Comment (velaja) [reply contrib/urls-maintenance.sh]: right i leave the cache there to speed up cleaning after tests

File: contrib/urls-maintenance.sh (hunk 3 above)
Comment (velaja) [reply contrib/urls-maintenance.sh]: the script is going posix compliant. at least i'm trying.

File: contrib/urls-maintenance.sh (hunk 1 above)
Comment (Minoru) [reply contrib/urls-maintenance.sh]: Ah, okay. I'll keep this comment unresolved for now, so we don't forget to address it before merging this.

**Comment by Minoru** [comment]
@velaja, are you still working on this?

Sorry I lost track of this PR. I think it's generally better to merge a "good enough" version, and make more PRs to improve it; otherwise we'll keep reviewing the same PR for months as it gets closer and closer to perfection.

Can you make a list of things you're trying to implement right now, and then we can agree on which ones should be part of this PR (probably fixes to review comments) and which ones can be separate PRs (probably new features)?

**Comment by Minoru** [comment]
Despite bugs in corner cases, this script is generally useful, so I'm merging this. @velaja (and everyone else), please submit more PRs to polish this. And @velaja —thank you once again for working on this!
